File: Easy/20. Valid Parenthesis.java
// 20. Valid Parenthesis
class Solution {
    public boolean isValid(String s) {
        int n = s.length();
        char[] stack = new char[n];
        int top = -1;

        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack[++top] = c;
            } else {
                if (top == -1)
                    return false;
                char topChar = stack[top--];
                if ((c == ')' && topChar != '(') || (c == '}' && topChar != '{') || (c == ']' && topChar != '[')) {
                    return false;
                }
            }
        }
        return top == -1;
    }
}

File: Easy/20. Valid Parenthesis.py
# 20. Valid Parentheses
class Solution:
    def isValid(self, s: str) -> bool:
        matching = {')': '(', '}': '{', ']': '['}
        stack = []

        for char in s:
            if char in matching.values():
                stack.append(char)
            elif char in matching.keys():
                if stack == [] or stack[-1] != matching[char]:
                    return False
                stack.pop()
        return stack == []
